### PR TITLE
BUG: Prevent uint64 overflow in Series.unique

### LINKED ISSUE #14721: Series.unique converts uint64 to int64 (with overflow)
```
>>> import pandas as pd
>>> import numpy as np
>>> s = pd.Series([1,2,2**63], dtype=np.dtype('uint64'))
>>> s
0                      1
1                      2
2    9223372036854775808
dtype: uint64
>>> s.unique()
array([                   1,                    2, -9223372036854775808])

```

Series.unique should preserve whatever type the values are.  But instead it changes from uint64 to int64.

INSTALLED VERSIONS
------------------
...
python: 3.4.3.final.0
python-bits: 64
...

...
pandas: 0.17.1
numpy: 1.10.4
...

### PR DESCRIPTION
**Pull request by gfyoung**
Introduces a `UInt64HashTable` class to hash `uint64` elements and prevent overflow in functions like `Series.unique`.

Closes #14721.

### CODE REVIEW

File: pandas/tests/test_base.py
Comment (jreback) [inline pandas/tests/test_base.py]: i think needs a bit more testing
most test are in algos for things like this

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jreback** [comment]
you will need to rebase after #14919. I moved a couple of things, and added a HashTable testing section.

Though what I would really like is to systematically add unit64 to as many places as we test other dtypes and see what breaks / falls out (can do in another PR). Then create issues for incompat places.

File: pandas/tests/test_base.py
Comment (gfyoung) [inline pandas/tests/test_base.py]: "Needs a bit more testing" is vague.  What sort of cases are you thinking of?  Secondly, the affected function has no relation to `algos`.  The code path is completely different.

However, you did catch a bug in `pd.unique`, which does travel through `algorithms.py` and can now benefit from my hashtable. :smile:

Comment (jreback) [inline pandas/tests/test_base.py]: not on the hashtable itself. ideally you should add uint64 to lots of places that we test int64 (construction type things). IOW, to actually exercise this code. I know this is vague. happy to merge this and to do a followup with more tests (and potentially breaking things). later.

Comment (gfyoung) [inline pandas/tests/test_base.py]: Okay, well that's a larger issue beyond this PR.  However, let's not merge until I patch `pd.unique`, which suffers from a similar issue as `Series.unique`.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
what was the bug (from before I moved unique)?

**Comment by gfyoung** [comment]
@jreback : The <a href="https://github.com/pandas-dev/pandas/pull/14915/files#diff-fafd7700903bc35b7e73b5e1fa65b6daR133">diff</a> explains it all.  We were smashing all integer types together into one hashtable.

**Comment by jreback** [comment]
ok, lgtm. ping on green.

**Comment by gfyoung** [comment]
@jreback : I changed my mind about changing `pd.unique` in this PR.  `core.algorithms` is riddled with `uint64` bugs that need to be checked.  I'll tackle in a follow-up.

**Comment by jreback** [comment]
> @jreback : Sounds good. I changed my mind about changing pd.unique in this PR. core.algorithms is riddled with uint64 bugs that need to be checked. I'll tackle in a follow-up.

totally fine. we can make a master issue of these if that is helpful.

**Comment by gfyoung** [comment]
@jreback : Everything is green, so ready to merge if there are no other concerns.

**Comment by jreback** [comment]
thanks!

as I said, ideally have an issue which greatly expands test coverage for uint64 (and just mark the tests that are failing (e.g. assert that they are failing or comment out)).

Then can come back around and fix.

**Pull request closed without merging**
